examples/rlhf/train_rlhf.py

```diff
@@ -0,0 +1,385 @@
+from copy import deepcopy
+
+import matplotlib.pyplot as plt
+import torch
+import torch.nn.functional as F
+
+from data import get_prompt_dataloader
+from models.actor_critic import init_actor_critic
+from models.reward import init_reward_model
+from tensordict import TensorDict
+from tensordict.nn import TensorDictModuleBase
+from torchrl.data.replay_buffers import (
+    LazyTensorStorage,
+    SamplerWithoutReplacement,
+    TensorDictReplayBuffer,
+)
+from torchrl.objectives import ClipPPOLoss
+from torchrl.objectives.value import GAE
+from tqdm import trange
+from transformers import GenerationConfig, GPT2Tokenizer
+from utils import get_file_logger, load_config
+
+EOS_TOKEN_ID = 50256
+
+
+class VmapCritic(TensorDictModuleBase):
+    def __init__(self, critic):
+        super().__init__()
+        self.in_keys = critic.in_keys
+        self.out_keys = critic.out_keys
+        self.module = critic
+
+    def forward(self, tensordict):
+        ndim = tensordict.ndim
+        training = self.module.training
+        self.module.eval()
+        td = torch.vmap(self.module, (ndim - 1,))(tensordict)
+        self.module.train(training)
+        # vmap sends this dim to the beginning so we need to send it back where it belongs
+        td = td.permute(*range(1, ndim), 0)
+        return tensordict.update(td)
+
+
+def logprobs_of_labels(logits, labels):
+    """Log probabilities of the labels
+
+    These are calculated from the logits."""
+    logprobs = F.log_softmax(logits, dim=-1)
+    logprobs_labels = torch.gather(logprobs, dim=-1, index=labels.unsqueeze(-1))
+    return logprobs_labels.squeeze(-1)
+
+
+@torch.no_grad()
+def generate(model, batch, ref_model, max_new_tokens=50):
```

Comment:
missing docstring
worth porting to repo?

Reply:
See comment on `log_probs_of_labels` above.